### PR TITLE
kissat: init at 2.0.1

### PR DESCRIPTION
**Pull request by shnarazk**
###### Motivation for this change

Kissat is one of the most advanced SAT solvers, getting great results in recent SAT competitions.
  - https://github.com/arminbiere/kissat
  - http://fmv.jku.at/kissat/
  - https://satcompetition.github.io/2020/results.html
  - https://satcompetition.github.io/2021/results.html

Please add it.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/science/logic/kissat/default.nix [hunk 1]
```diff
@@ -0,0 +1,55 @@
+{ lib, stdenv, fetchFromGitHub, drat-trim }:
+
+stdenv.mkDerivation rec {
+  pname = "kissat";
+  version = "2.0.1-abfa45f";
```
Comment (erdnaxe) [inline pkgs/applications/science/logic/kissat/default.nix:5]: Will `2.0.1` always point to this commit? If yes, maybe:
```suggestion
  version = "2.0.1";
```

(I cannot make my mind if https://nixos.org/manual/nixpkgs/unstable/#sec-package-naming allows doing `2.0.1-abfa45f` or not)

**Review by erdnaxe — COMMENTED** [review]
Result of `nixpkgs-review pr 139294` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>kissat</li>
  </ul>
</details>

File: pkgs/applications/science/logic/kissat/default.nix (hunk 1 above)
Comment (shnarazk) [reply pkgs/applications/science/logic/kissat/default.nix:5]: There's no official (or tag) release yet. I can't guarantee but I think this commit is the only one for 2.0.1. So I'd like to take 'kissat-2.0.1'.

File: pkgs/applications/science/logic/kissat/default.nix [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+  # the configure script is not generated by autotools and does not accept the
+  # arguments that the default configurePhase passes like --prefix and --libdir
+  configurePhase = ''
+    runHook preConfigure
+
+    ./configure
+
+    runHook postConfigure
+  '';
```
Comment (hrhino) [inline pkgs/applications/science/logic/kissat/default.nix:31]:
```suggestion
  dontAddPrefix = true;
  setOutputFlags = false;
```

File: pkgs/applications/science/logic/kissat/default.nix [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  };
+
+  buildInputs = [ drat-trim ];
+  outputs = [ "out" "dev" "lib" ];
+  doCheck = true;
+
+  # 'make test' assumes that /etc/passwd is not writable.
+  prePatch = ''
+    patch -p1 -i ${./writable-passwd-is-ok.patch}
+  '';
```
Comment (hrhino) [inline pkgs/applications/science/logic/kissat/default.nix:21]:
```suggestion
  patches = [ ./writable-passwd-is-ok.patch ];
    patch -p1 -i ${./writable-passwd-is-ok.patch}
  '';
```

File: pkgs/applications/science/logic/kissat/default.nix [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  src = fetchFromGitHub {
+    owner = "arminbiere";
+    repo = "kissat";
+    rev = "abfa45fb782fa3b7c6e2eb6b939febe74d7270b7";
+    sha256 = "06pbmkjxgf2idhsrd1yzvbxr2wf8l06pjb38bzbygm6n9ami89b8";
+  };
+
+  buildInputs = [ drat-trim ];
+  outputs = [ "out" "dev" "lib" ];
+  doCheck = true;
```
Comment (hrhino) [inline pkgs/applications/science/logic/kissat/default.nix:16]:
```suggestion
  outputs = [ "out" "dev" "lib" ];
  
  checkInputs = [ drat-trim p7zip ];
  doCheck = true;
```

File: pkgs/applications/science/logic/kissat/default.nix [hunk 5]
```diff
@@ -0,0 +1,55 @@
+{ lib, stdenv, fetchFromGitHub, drat-trim }:
```
Comment (hrhino) [inline pkgs/applications/science/logic/kissat/default.nix:1]:
```suggestion
{ lib, stdenv, fetchFromGitHub
, drat-trim, p7zip
}:
```

**Review by hrhino — CHANGES_REQUESTED** [review]
As far as I can tell, `drat-trim` is only needed for the tests.
Also, the test output complains about having to skip some tests because of missing `7z` and also because of something called `drabt` missing, which I can't figure out.

File: pkgs/applications/science/logic/kissat/default.nix [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+stdenv.mkDerivation rec {
+  pname = "kissat";
+  version = "2.0.1";
+
+  src = fetchFromGitHub {
+    owner = "arminbiere";
+    repo = "kissat";
+    rev = "abfa45fb782fa3b7c6e2eb6b939febe74d7270b7";
```
Comment (SuperSandro2000) [inline pkgs/applications/science/logic/kissat/default.nix:10]:
```suggestion
    # https://github.com/arminbiere/kissat/issues/18
    rev = "abfa45fb782fa3b7c6e2eb6b939febe74d7270b7";
```

**Comment by shnarazk** [comment]
I set `checkInput` with drat-trim and p7.
We can ignore the message about drabt, because either drat-trim or drabt is required to check all tests. Both are proof checkers.

**Comment by hrhino** [comment]
Result of `nixpkgs-review pr 139294` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>kissat</li>
  </ul>
</details>

**Review by hrhino — APPROVED** [review]

**Comment by hrhino** [comment]
LGTM; I tried it out and it seems to work as expected

File: pkgs/applications/science/logic/kissat/writable-passwd-is-ok.patch [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+--- a/test/testfile.c
++++ b/test/testfile.c
+@@ -92,8 +92,8 @@ do { \
+   WRITABLE (true, "../test/file/non-existing");
+   WRITABLE (false, "/kissat-test-file-writable");
+   WRITABLE (false, "non-existing-directory/file-in-non-existing-directory");
+-  if (kissat_file_exists ("/etc/passwd"))
+-    WRITABLE (false, "/etc/passwd");
++//  if (kissat_file_exists ("/etc/passwd"))
++//    WRITABLE (false, "/etc/passwd");
```
Comment (SuperSandro2000) [inline pkgs/applications/science/logic/kissat/writable-passwd-is-ok.patch:12]:
```suggestion
```

File: pkgs/applications/science/logic/kissat/default.nix [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    mkdir -p "$out/share/doc/${pname}/"
+    install -Dm0644 {LICEN?E,README*,VERSION} "$out/share/doc/${pname}/"
+
+    runHook postInstall
+  '';
+
+  meta = with lib; {
+    description = "A 'keep it simple and clean bare metal SAT solver' written in C";
+    longDescription = ''
+      Kissat is a "keep it simple and clean bare metal SAT solver" written in C. It is a port of CaDiCaL back to C with improved data structures, better scheduling of inprocessing and optimized algorithms and implementation.
```
Comment (SuperSandro2000) [inline pkgs/applications/science/logic/kissat/default.nix:45]:
```suggestion
      Kissat is a "keep it simple and clean bare metal SAT solver" written in C.
      It is a port of CaDiCaL back to C with improved data structures,
      better scheduling of inprocessing and optimized algorithms and implementation.
```

File: pkgs/applications/science/logic/kissat/default.nix [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+  setOutputFlags = false;
+
+  installPhase = ''
+    runHook preInstall
+
+    install -Dm0755 build/kissat "$out/bin/kissat"
+    install -Dm0644 src/kissat.h "$dev/include/kissat.h"
+    install -Dm0644 build/libkissat.a "$lib/lib/libkissat.a"
+    mkdir -p "$out/share/doc/${pname}/"
+    install -Dm0644 {LICEN?E,README*,VERSION} "$out/share/doc/${pname}/"
```
Comment (SuperSandro2000) [inline pkgs/applications/science/logic/kissat/default.nix:37]:
```suggestion
    install -dDm0644 {LICEN?E,README*,VERSION} "$out/share/doc/${pname}/"
```

Comment (shnarazk) [reply pkgs/applications/science/logic/kissat/default.nix]: It does not work.
```
make[1]: Leaving directory '/private/tmp/nix-build-kissat-2.0.1.drv-0/source/build'
installing
install: cannot create directory 'LICENSE': File exists
install: cannot create directory 'README.md': File exists
install: cannot create directory 'VERSION': File exists
builder for '/nix/store/2w6cry2p45a4l7as408x2mfjcacy6j05-kissat-2.0.1.drv' failed with exit code 1
```

**Comment by smancill** [comment]
Result of `nixpkgs-review pr 139294` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>kissat</li>
  </ul>
</details>

**Pull request merged**